File: leetcode/go/1143.longest-common-subsequence.go
func longestCommonSubsequence(text1 string, text2 string) int {
    m, n := len(text1), len(text2)
    dp := make([][]int, m+1)
    for i := range m+1 {
        dp[i] = make([]int, n+1)
    }
    for i := 0; i < m ; i++ {
        for j := 0; j < n ; j++ {
            if text1[i] == text2[j] {
                dp[i+1][j+1] = 1 + dp[i][j]
            } else {
                dp[i+1][j+1] = Max(dp[i+1][j], dp[i][j+1])
            }
        }
    }
    return dp[m][n]
}

func Max(a int, b int) int {
    if a > b {return a}
    return b
}

File: leetcode/rust-leetcode/src/1143.longest-common-subsequence.rs
use std::cmp::max;
impl Solution {
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
        let (m, n) = (text1.len(), text2.len());
        let mut dp = vec![vec![0; n+1]; m+1];
        let t1 = text1.into_bytes();
        let t2 = text2.into_bytes();
        for i in 0..m {
            for j in 0..n {
                dp[i+1][j+1] = if t1[i] == t2[j] {
                    1 + dp[i][j]
                } else {
                    max(dp[i][j+1], dp[i+1][j])
                }
            }
        }
        dp[m][n]
    }
}

